fix(ngcc): cope with packages following APF v14+

gkalpak opened the pull request:
In PR #45405, the Angular Package Format (APF) was updated so that secondary entry-points (such as `@angular/common/http`) do not have their own `package.json` file, as they used to. Instead, the paths to their various formats and types are exposed via the primary `package.json` file's `exports` property. As an example, see the v13 [@angular/common/http/package.json][1] and compare it with the v14 [@angular/common/package.json > exports][2].

Previously, `ngcc` was not able to analyze such v14+ entry-points and would instead error as it considered such entry-points missing.

This commit addresses the issue by detecting this situation and synthesizing a `package.json` file for the secondary entry-points based on the `exports` property of the primary `package.json` file. This data is only used by `ngcc` in order to determine that the entry-point does not need further processing, since it is already in Ivy format.

##
FYI, I have tested this on `ngcc-validation` and it makes CI green: [CI run][3]

[1]: https://unpkg.com/browse/@angular/common@13.3.5/http/package.json
[2]: https://unpkg.com/browse/@angular/common@14.0.0-next.15/package.json
[3]: https://app.circleci.com/pipelines/github/angular/ngcc-validation/18407/workflows/36ae9514-3d96-4407-b104-a372a634e17c

mary-poppins commented:
You can preview cda0f67 at https://pr45833-cda0f67.ngbuilds.io/.

mary-poppins commented:
You can preview c92207d at https://pr45833-c92207d.ngbuilds.io/.

mary-poppins commented:
You can preview 330569e at https://pr45833-330569e.ngbuilds.io/.

mary-poppins commented:
You can preview 7c75b0b at https://pr45833-7c75b0b.ngbuilds.io/.

mary-poppins commented:
You can preview 1ab8b94 at https://pr45833-1ab8b94.ngbuilds.io/.

mary-poppins commented:
You can preview d4c3a78 at https://pr45833-d4c3a78.ngbuilds.io/.

JoostK commented on packages/compiler-cli/ngcc/src/dependencies/module_resolver.ts:
```diff
@@ -136,9 +136,25 @@ export class ModuleResolver {
    *
    * This is achieved by checking for the existence of `${modulePath}/package.json`.
+   * If there is no `package.json`, we check whether this is an APF v14+ secondary entry-point,
+   * which does not have each own `package.json` but has an `exports` entry in the package's primary
```
should this be

```suggestion
   * which does not have its own `package.json` but has an `exports` entry in the package's primary
```

gkalpak replied:
😱 
Fixed.

JoostK commented on packages/compiler-cli/ngcc/src/utils.ts:
```diff
@@ -186,3 +186,33 @@ export function loadJson<T extends JsonObject = JsonObject>(
+
+  // Read the data (if it exists).
+  const exportsData =
+      (primaryPackageJson.exports as JsonObject)[exportsKey] as JsonObject | undefined;
```
[NodeJS export maps](https://nodejs.org/api/packages.html#exports) can also be `string | string[]`, and each individual subpath can also be just `string` (for non-conditional exports). APF may always contain conditional export maps, but it wouldn't hurt to verify these casts?

JoostK commented on packages/compiler-cli/ngcc/test/utils_spec.ts:
```diff
@@ -276,3 +276,58 @@ runInEachFileSystem(() => {
+  beforeEach(() => fs = getFileSystem());
+
+  describe('loadSecondaryEntryPointInfoForApfV14()', () => {
+    it('should return `null` of the primary `package.json` failed to be loaded', () => {
```
```suggestion
    it('should return `null` if the primary `package.json` failed to be loaded', () => {
```

JoostK reviewed (commented):
Nice work!

gkalpak committed:
fix(ngcc): cope with packages following APF v14+

In PR #45405, the Angular Package Format (APF) was updated so that
secondary entry-points (such as `@angular/common/http`) do not have
their own `package.json` file, as they used to. Instead, the paths to
their various formats and types are exposed via the primary
`package.json` file's `exports` property. As an example, see the v13
[@angular/common/http/package.json][1] and compare it with the v14
[@angular/common/package.json > exports][2].

Previously, `ngcc` was not able to analyze such v14+ entry-points and
would instead error as it considered such entry-points missing.

This commit addresses the issue by detecting this situation and
synthesizing a `package.json` file for the secondary entry-points based
on the `exports` property of the primary `package.json` file. This data
is only used by `ngcc` in order to determine that the entry-point does
not need further processing, since it is already in Ivy format.

[1]: https://unpkg.com/browse/@angular/common@13.3.5/http/package.json
[2]: https://unpkg.com/browse/@angular/common@14.0.0-next.15/package.json

gkalpak committed:
refactor(ngcc): remove unused dependencies

Remove some unused dependencies from `ngcc` source code.

gkalpak committed:
refactor(ngcc): move `loadPackageJson()` and related types to `utils.ts`

Move the `loadPackageJson()` helper (and associated generic types, such
as `JsonObject`) from `packages/entry_point.ts` to `utils.ts` and also
rename it to `loadJson()`. This way, they can be used in other places in
future commits, without introducing cyclical dependencies.

gkalpak committed:
fixup! fix(ngcc): cope with packages following APF v14+

JoostK reviewed (approved):
LGTM

mary-poppins commented:
You can preview 3686a06 at https://pr45833-3686a06.ngbuilds.io/.

AndrewKushnir commented:
This PR was merged into the repository by commit d4e949f45c35796f1c71f7dd3bd705293984aafd.
